metpy/cbook.py

```diff
@@ -39,4 +37,103 @@ def get_test_data(fname, as_file_obj=True):
+            # a version file with the current cache version
+            self.cache_home = self.get_cache_location()
+            with open(cache_version_file_path, 'w') as f:
+                f.write('{}\n'.format(self.cache_version))
+
+    def get_cache_location(self):
+        """Get the cache location."""
+
+        # Look for the staticdata directory (i.e. this is a git checkout)
+        if os.path.exists(os.path.join(os.path.dirname(__file__), '..', 'staticdata')):
+            cache_path = os.path.join(os.path.dirname(__file__), '..', 'staticdata')
+
+        # If there is no static data directory, use the OS default
+        else:
+            # Get the cache location for this OS
+            cache_path = self.get_os_cache_path()
+
+        # Look for a user-defined cache location - this takes priority over anything else
+        if os.environ.get('METPY_CACHE_LOCATION'):
+            cache_path = os.environ.get('METPY_CACHE_LOCATION')
+
+        # Make sure we expand ~ to the user home
+        cache_path = os.path.expanduser(cache_path)
+
+        # Make sure there is a cache directory there. If not, let's create it so we have
+        # somewhere to put the data.
+        if not os.path.exists(cache_path):
+            os.makedirs(cache_path)
+
+        return cache_path
+
+    def get_os_cache_path(self):
+        platform = sys.platform
+
+        if platform == 'darwin':
+            cache_path = '~/Library/Caches/MetPy'
+        elif platform == 'win32':
+            cache_path = '~/AppData/Local/MetPy/cache'
+        else:  # *NIX
+            cache_path = '~/.cache/MetPy'
+
+        return os.path.expanduser(cache_path)
+
+    def get_data(self, fname, as_file_obj=True):
+        """Get data from the cache."""
+
+        file_path = os.path.join(self.cache_home, fname)
+
+        # See if we already have the file locally
+        if os.path.exists(file_path):
+            return file_path
+
+        # If we don't have it, try to download it
+        else:
+            self.download_data(fname)
+
+        return file_path
+
+    def download_data(self, fname):
+        """Download remote data to the cache."""
+        if '.dirty' in __version__:
+            gh_tag = 'master'
+        else:
+            gh_tag = 'v{}'.format(__version__)
+        github_data_path = 'https://github.com/Unidata/MetPy/raw/{}/testdata'.format(gh_tag)
+
+        cache_fname = os.path.join(self.cache_home, fname)
+        cache_subpath = os.path.split(cache_fname)[0]
+
+        # Make sure if it is a subdirectory that it exists before pulling down data
+        if not os.path.exists(cache_subpath):
+            os.makedirs(cache_subpath)
+
+        # TODO: Error handling if this isn't a valid remote file?
+        urllib.request.urlretrieve(github_data_path + '/' + fname, cache_fname)
+
+    def clear_cache(self):
+        """Remove all cached data for MetPy."""
+        shutil.rmtree(self.cache_home)
```

Comment:
For testing, we can mock out `shultil.rmtree` and just check that it's called with an acceptable parameter.